src/ProjectOrigin.VerifiableEventStore/Services/BatchProcessor/BatchProcessorJob.cs

```diff
@@ -1,47 +1,53 @@
+using ProjectOrigin.VerifiableEventStore.Models;
 using ProjectOrigin.VerifiableEventStore.Services.BlockchainConnector;
 using ProjectOrigin.VerifiableEventStore.Services.EventStore;
+using ProjectOrigin.VerifiableEventStore.Services.TransactionStatusCache;
 
 namespace ProjectOrigin.VerifiableEventStore.Services.BatchProcessor;
 
 public sealed class BatchProcessorJob
 {
-    private const int NumberOfBatches = 10;
-
     private readonly IBlockchainConnector _blockchainConnector;
     private readonly IEventStore _eventStore;
-    private readonly TimeSpan _period = TimeSpan.FromSeconds(30);
+    private readonly ITransactionStatusService _statusService;
+    private readonly TimeSpan _period = TimeSpan.FromSeconds(5);
 
-    public BatchProcessorJob(IBlockchainConnector blockchainConnector, IEventStore eventStore)
+    public BatchProcessorJob(IBlockchainConnector blockchainConnector, IEventStore eventStore, ITransactionStatusService statusService)
     {
         _blockchainConnector = blockchainConnector;
         _eventStore = eventStore;
+        _statusService = statusService;
     }
 
     public async Task Execute(CancellationToken stoppingToken)
     {
-        // go check if we have any batches that are full
-        // If we have any - then go get the batch and the events
-        var batches = await _eventStore.GetBatchesForFinalization(NumberOfBatches);
-        foreach (var item in batches)
+        // As long as there is batches to finalize continue finalizing.
+        while (await _eventStore.TryGetNextBatchForFinalization(out var batch))
         {
-            var events = await _eventStore.GetEventsForBatch(item);
+            var events = await _eventStore.GetEventsForBatch(batch.Id);
             var merkleRoot = events.CalculateMerkleRoot(x => x.Content);
 
-            var transaction = await _blockchainConnector.PublishBytes(merkleRoot);
+            var blockchainTransaction = await _blockchainConnector.PublishBytes(merkleRoot);
 
             // This should be in another service
-            var block = await _blockchainConnector.GetBlock(transaction);
+            var block = await _blockchainConnector.GetBlock(blockchainTransaction);
             while (block == null || !block.Final)
             {
                 await Task.Delay(_period, stoppingToken);
-                block = await _blockchainConnector.GetBlock(transaction);
+                block = await _blockchainConnector.GetBlock(blockchainTransaction);
             }
 
             // Update batch in EventStore
-            await _eventStore.FinalizeBatch(item, block.BlockId, transaction.TransactionHash);
-        }
+            await _eventStore.FinalizeBatch(batch.Id, block.BlockId, blockchainTransaction.TransactionHash);
 
-        return;
+            foreach (var e in events)
+            {
+                await _statusService.SetTransactionStatus(e.TransactionId, new Models.TransactionStatusRecord(TransactionStatus.Committed));
```

Comment:
The transaction status service could take a list/an array of transactions and set them to Committed.